### PR TITLE
OIDC: Render publisher on each release in admin view

### PR DESCRIPTION
**Pull request by woodruffw**
~~Needs tests.~~

~~This required some more involved changes: each `Release` now tracks its OIDC publisher (if present) via `Release.oidc_publisher`, and I've added `request.oidc_publisher` as a shim to reduce the amount of inline `isinstance` checks necessary.~~

This adds two new `File` columns:

* `File.uploading_user`: the `User` who uploaded this `File`, if uploaded by a user identity;
* `File.oidc_publisher`: the `OIDCPublisher` who uploaded this `File`, if uploaded by an OIDC identity;

This in turn allows us to display each `File`'s upload information in the admin pane, including if it came from an OIDC publisher.

Historical `File`s are not backloaded.

### CODE REVIEW

File: warehouse/oidc/models.py [hunk 1]
```diff
@@ -194,6 +194,12 @@ class OIDCPublisher(OIDCPublisherMixin, db.Model):
     macaroons = orm.relationship(
         Macaroon, backref="oidc_publisher", cascade="all, delete-orphan", lazy=True
     )
+    files = orm.relationship(
+        File,
+        backref="oidc_publisher",
+        cascade="all, delete-orphan",
```
Comment (woodruffw) [inline warehouse/oidc/models.py:200]: Flagging for review: not sure this cascade is correct.

Comment (di) [reply warehouse/oidc/models.py:200]: I believe this would delete all `Files` when the `OIDCPublisher` was deleted, so no :)

**Review by ewdurbin — COMMENTED** [review]
Overall I dig this. I am pondering if we should include backfill here and add a constraints like:

```
CheckConstraint('NOT(oidc_publisher_id IS NULL AND uploading_user_id IS NULL)')
CheckConstraint('NOT(oidc_publisher_id IS NOT NULL AND uploading_user_id IS NOT NULL)')
```

If they're mutually exclusive it's always best to enforce that in the db.

**Comment by woodruffw** [comment]
Adding those constraints makes sense to me! I can do that now.

I think a backfill also makes sense, although I wonder about potentially being misleading: it's possible that existing projects have releases created by one user, but release files uploaded by multiple. That isn't recorded anywhere at the moment, but the backfill will make it appear as if the release-creating user is responsible for all files.

That's a minor issue IMO, though 🙂

**Comment by ewdurbin** [comment]
Theoretically we can do a best effort by finding the upload in the Journal?

**Comment by ewdurbin** [comment]
@woodruffw I'm also OK seeing this go without constraints initially, followed by PR for backfill migration + constraint enforcement.

**Comment by di** [comment]
Constraints make sense, but I think backfill is out of scope for this particular thing -- we should file an issue to address that.

**Comment by ewdurbin** [comment]
I don't think we can have the constraint without the backfill?

**Comment by woodruffw** [comment]
> I don't think we can have the constraint without the backfill?

Yeah, these would need to happen together, I think.

**Comment by ewdurbin** [comment]
Ah, solid yeah. Let's see the "you can only set one" constraint in this PR and at least file the issue for backfill + "you must set one" constraint.

**Comment by woodruffw** [comment]
> Ah, solid yeah. Let's see the "you can only set one" constraint in this PR and at least file the issue for backfill + "you must set one" constraint.

SGTM!

**Comment by woodruffw** [comment]
Not needed for coverage, but I went ahead and added some unit tests for the new constraint as well.

**Comment by di** [comment]
Chatted with William: I'm concerned about publisher details about files going away if publishers are deleted. We're going to rework this to use the corresponding security event instead, which should contain most of the detail we need here (now and in the future).

**Comment by woodruffw** [comment]
Superseded by #13205, closing.

**Pull request closed without merging**